vm/sdk/src/wallet.rs

```diff
@@ -0,0 +1,61 @@
+//! Definitions of the reference wallet interface.
+
+use athena_interface::Address;
+use parity_scale_codec::{Decode, Encode};
+
+use crate::Pubkey;
+
+#[derive(Clone, Copy, Debug, Default, Encode, Decode, PartialEq, Eq)]
+pub struct SpendArguments {
+  pub recipient: Address,
+  pub amount: u64,
+}
+
+pub trait WalletProgram {
+  fn spawn(owner: Pubkey) -> Address;
+  fn spend(&self, args: SpendArguments);
+  fn proxy(&self, destination: Address, args: &[u8]);
+  fn deploy(&self, code: Vec<u8>) -> Address;
+}
+
+pub fn encode_spend(state: Vec<u8>, args: SpendArguments) -> Vec<u8> {
+  let mut encoded = state;
+  encoded.extend(args.encode());
+  encoded
+}
+
+pub fn encode_spawn(pubkey: Pubkey) -> Vec<u8> {
+  pubkey.encode()
+}
+
+pub fn encode_deploy(state: Vec<u8>, code: Vec<u8>) -> Vec<u8> {
+  let mut encoded = state;
+  encoded.extend(code.encode());
+  encoded
+}
+
+#[cfg(test)]
+mod tests {
+  use parity_scale_codec::{Decode, Encode, IoReader};
+
+  use super::SpendArguments;
+
+  #[test]
+  fn encode_decode_spend() {
+    let wallet = Vec::<i32>::from([1, 2, 3, 4]);
```

Comment:
I'm confused, why is this a valid wallet state? Why not actually instantiate the wallet here and encode the instance?

Reply:
It's a dummy state for testing - it doesn't need to be valid for the sake of this test.

> Why not actually instantiate the wallet here and encode the instance?

For simplicity and to avoid cyclic dependencies. The `struct Wallet` we have in examples/wallet/program, which implements `WalletProgram`, is only one of the possible implementations (and it already imports this pkg so trying to import it here would cause a cyclic dep).

Reply:
I just find this test kinda counterintuitive and hard to read, e.g.:

```
let decoded_wallet = Vec::<i32>::decode(&mut input_reader).unwrap();
```

This isn't a "decoded wallet" at all, it's a vector of arbitrary integers.

I think it could be made more readable with some explanatory comments.

Reply:
> This isn't a "decoded wallet" at all, it's a vector of arbitrary integers.

A wallet structure could be anything :shrug:. This:
```rust
struct Foo(Vec<i32>);
impl WalletProgram for Foo {
...
}
```

is a valid wallet, which state consists of a vector of i32s. 

I can change it to a more "natural"  `struct Wallet {...}` if this is confusing.

Reply:
Changed to a Wallet struct for clarity.